{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e78f16e1",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [],
   "source": [
    "from IPython.display import Markdown, display\n",
    "\n",
    "date = '2022-03-08'\n",
    "practice = '../_practice/' + date + '.md'\n",
    "prepare =  '../_prepare/' + date + '.md'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66b28c9f",
   "metadata": {},
   "source": [
    "# How can I work on a remote server?\n",
    "\n",
    "Today we will connect to a remote server, use an HPC system with scheduler, and learn new bash commands for working with the *content* of files.\n",
    "\n",
    "\n",
    "\n",
    "## What are remote servers and HPC systems?\n",
    "\n",
    "![diagram illustrating a remot connection to a login node and compute cluster](../img/hpcsketch.svg)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "## Connecting to Seawulf\n",
    "\n",
    "We connect with secure shell or `ssh` from our terminal (GitBash or Putty on windows) to URI's teaching High Performance Computing (HPC) Cluster [Seawulf](https://web.uri.edu/hpc-research-computing/using-seawulf/).  \n",
    "\n",
    "````{margin}\n",
    "```{warning}\n",
    "This cluster is for course related purposes at URI, if you want to use a HPC system of some sort for a side project, consider Amazon Web Services, Google Cloud, or Microsoft Azul services, you can get some allocation for free a a student.  \n",
    "\n",
    "If you are doing research supervised by a URI professor, there are other servers on campus and URI participates in a regional HPC resource as well.\n",
    "```\n",
    "````\n",
    "\n",
    "Our login is the part of your uri e-mail address before the @ and I will tell you how to find your default password if you missed class.\n",
    "```\n",
    "ssh -l brownsarahm seawulf.uri.edu\n",
    "```\n",
    "\n",
    "When it logs in it looks like this and requires you to change your password.  They configure it with a default and with it past expired.\n",
    "```\n",
    "The authenticity of host 'seawulf.uri.edu (131.128.217.210)' can't be established.\n",
    "ECDSA key fingerprint is SHA256:RwhTUyjWLqwohXiRw+tYlTiJEbqX2n/drCpkIwQVCro.\n",
    "Are you sure you want to continue connecting (yes/no/[fingerprint])? y\n",
    "Please type 'yes', 'no' or the fingerprint: yes\n",
    "Warning: Permanently added 'seawulf.uri.edu,131.128.217.210' (ECDSA) to the list of known hosts.\n",
    "brownsarahm@seawulf.uri.edu's password:\n",
    "You are required to change your password immediately (root enforced)\n",
    "WARNING: Your password has expired.\n",
    "You must change your password now and login again!\n",
    "Changing password for user brownsarahm.\n",
    "Changing password for brownsarahm.\n",
    "(current) UNIX password:\n",
    "New password:\n",
    "Retype new password:\n",
    "passwd: all authentication tokens updated successfully.\n",
    "Connection to seawulf.uri.edu closed.\n",
    "```\n",
    "after you give it a new password, then it logs you out and you have to log back in.\n",
    "\n",
    "```\n",
    "brownsarahm@~ $ ssh -l brownsarahm seawulf.uri.edu\n",
    "```\n",
    "\n",
    "\n",
    "```\n",
    "brownsarahm@seawulf.uri.edu's password:\n",
    "Last login: Tue Mar  8 12:52:38 2022 from 172.20.133.152\n",
    "```\n",
    "\n",
    "We have logged into our home directory which is empty\n",
    "\n",
    "```\n",
    "[brownsarahm@seawulf ~]$ ls\n",
    "```\n",
    "\n",
    "```\n",
    "[brownsarahm@seawulf ~]$ pwd\n",
    "/home/brownsarahm\n",
    "```\n",
    "\n",
    "```\n",
    "[brownsarahm@seawulf ~]$ whoami\n",
    "brownsarahm\n",
    "```\n",
    "\n",
    "Notice that the prompt says `uriusername@seawulf` to indicate that you are logged into the server, not working locally.  \n",
    "\n",
    "## Downloading files\n",
    "\n",
    "```\n",
    "[brownsarahm@seawulf ~]$ wget http://www.hpc-carpentry.org/hpc-shell/files/bash-lesson.tar.gz\n",
    "--2022-03-08 12:58:09--  http://www.hpc-carpentry.org/hpc-shell/files/bash-lesson.tar.gz\n",
    "Resolving www.hpc-carpentry.org (www.hpc-carpentry.org)... 104.21.33.152, 172.67.146.136\n",
    "Connecting to www.hpc-carpentry.org (www.hpc-carpentry.org)|104.21.33.152|:80... connected.\n",
    "HTTP request sent, awaiting response... 200 OK\n",
    "Length: 12534006 (12M) [application/gzip]\n",
    "Saving to: ‘bash-lesson.tar.gz’\n",
    "\n",
    "100%[======================================>] 12,534,006  4.19MB/s   in 2.9s   \n",
    "\n",
    "2022-03-08 12:58:12 (4.19 MB/s) - ‘bash-lesson.tar.gz’ saved [12534006/12534006]\n",
    "\n",
    "```\n",
    "\n",
    "```\n",
    "[brownsarahm@seawulf ~]$ ls\n",
    "bash-lesson.tar.gz\n",
    "```\n",
    "\n",
    "```\n",
    "[brownsarahm@seawulf ~]$ tar -xvf bash-lesson.tar.gz\n",
    "dmel-all-r6.19.gtf\n",
    "dmel_unique_protein_isoforms_fb_2016_01.tsv\n",
    "gene_association.fb\n",
    "SRR307023_1.fastq\n",
    "SRR307023_2.fastq\n",
    "SRR307024_1.fastq\n",
    "SRR307024_2.fastq\n",
    "SRR307025_1.fastq\n",
    "SRR307025_2.fastq\n",
    "SRR307026_1.fastq\n",
    "SRR307026_2.fastq\n",
    "SRR307027_1.fastq\n",
    "SRR307027_2.fastq\n",
    "SRR307028_1.fastq\n",
    "SRR307028_2.fastq\n",
    "SRR307029_1.fastq\n",
    "SRR307029_2.fastq\n",
    "SRR307030_1.fastq\n",
    "SRR307030_2.fastq\n",
    "```\n",
    "\n",
    "```\n",
    "[brownsarahm@seawulf ~]$ ls\n",
    "bash-lesson.tar.gz                           SRR307026_1.fastq\n",
    "dmel-all-r6.19.gtf                           SRR307026_2.fastq\n",
    "dmel_unique_protein_isoforms_fb_2016_01.tsv  SRR307027_1.fastq\n",
    "gene_association.fb                          SRR307027_2.fastq\n",
    "SRR307023_1.fastq                            SRR307028_1.fastq\n",
    "SRR307023_2.fastq                            SRR307028_2.fastq\n",
    "SRR307024_1.fastq                            SRR307029_1.fastq\n",
    "SRR307024_2.fastq                            SRR307029_2.fastq\n",
    "SRR307025_1.fastq                            SRR307030_1.fastq\n",
    "SRR307025_2.fastq                            SRR307030_2.fastq\n",
    "```\n",
    "\n",
    "## Working with large files\n",
    "\n",
    "One of these files, contains the entire genome for the common fruitfly, let's take a look at it: \\\n",
    "\n",
    "```\n",
    "[brownsarahm@seawulf ~]$ cat dmel-all-r6.19.gtf\n",
    "```\n",
    "\n",
    "```\n",
    "X\tFlyBase\tgene\t19961297\t19969323\t.\t+\t.\tgene_id \"FBgn0031081\"; gene_symbol \"Nep3\";\n",
    "2L\tFlyBase\t3UTR\t782825\t782885\t.\t+\t.\tgene_id \"FBgn0041250\"; gene_symbol \"Gr21a\"; transcript_id \"FBtr0331651\"; transcript_symbol \"Gr21a-RB\";\n",
    "```\n",
    "```{warning}\n",
    "this output is truncated for display purposes\n",
    "```\n",
    "\n",
    "We see that this actually take a long time to output and is way tooo much information to actually read.  In fact, in order to make the website work, I had to cut that content using command line tools, my text editor couldn't open the file and GitHub was unhappy when I pushed it.\n",
    "\n",
    "\n",
    "For a file like this, we don't really want to read the whole file but we do need to know what it's strucutred like in order to design programs to work with it.  \n",
    "\n",
    "`head` lets us look at the first 10 lines.\n",
    "\n",
    "```\n",
    "[brownsarahm@seawulf ~]$ head dmel-all-r6.19.gtf\n",
    "```\n",
    "\n",
    "```\n",
    "X\tFlyBase\tgene\t19961297\t19969323\t.\t+\t.\tgene_id \"FBgn0031081\"; gene_symbol \"Nep3\";\n",
    "X\tFlyBase\tmRNA\t19961689\t19968479\t.\t+\t.\tgene_id \"FBgn0031081\"; gene_symbol \"Nep3\"; transcript_id \"FBtr0070000\"; transcript_symbol \"Nep3-RA\";\n",
    "X\tFlyBase\t5UTR\t19961689\t19961845\t.\t+\t.\tgene_id \"FBgn0031081\"; gene_symbol \"Nep3\"; transcript_id \"FBtr0070000\"; transcript_symbol \"Nep3-RA\";\n",
    "X\tFlyBase\texon\t19961689\t19961845\t.\t+\t.\tgene_id \"FBgn0031081\"; gene_symbol \"Nep3\"; transcript_id \"FBtr0070000\"; transcript_symbol \"Nep3-RA\";\n",
    "X\tFlyBase\texon\t19963955\t19964071\t.\t+\t.\tgene_id \"FBgn0031081\"; gene_symbol \"Nep3\"; transcript_id \"FBtr0070000\"; transcript_symbol \"Nep3-RA\";\n",
    "X\tFlyBase\texon\t19964782\t19964944\t.\t+\t.\tgene_id \"FBgn0031081\"; gene_symbol \"Nep3\"; transcript_id \"FBtr0070000\"; transcript_symbol \"Nep3-RA\";\n",
    "X\tFlyBase\texon\t19965006\t19965126\t.\t+\t.\tgene_id \"FBgn0031081\"; gene_symbol \"Nep3\"; transcript_id \"FBtr0070000\"; transcript_symbol \"Nep3-RA\";\n",
    "X\tFlyBase\texon\t19965197\t19965511\t.\t+\t.\tgene_id \"FBgn0031081\"; gene_symbol \"Nep3\"; transcript_id \"FBtr0070000\"; transcript_symbol \"Nep3-RA\";\n",
    "X\tFlyBase\texon\t19965577\t19966071\t.\t+\t.\tgene_id \"FBgn0031081\"; gene_symbol \"Nep3\"; transcript_id \"FBtr0070000\"; transcript_symbol \"Nep3-RA\";\n",
    "X\tFlyBase\texon\t19966183\t19967012\t.\t+\t.\tgene_id \"FBgn0031081\"; gene_symbol \"Nep3\"; transcript_id \"FBtr0070000\"; transcript_symbol \"Nep3-RA\";\n",
    "```\n",
    "\n",
    "We can use the `-n` parameter to change the number.\n",
    "\n",
    "And, tails shows the last few.\n",
    "\n",
    "```\n",
    "[brownsarahm@seawulf ~]$ tail dmel-all-r6.19.gtf\n",
    "```\n",
    "\n",
    "which in this case looks mostly the same\n",
    "```\n",
    "2L\tFlyBase\texon\t782124\t782181\t.\t+\t.\tgene_id \"FBgn0041250\"; gene_symbol \"Gr21a\"; transcript_id \"FBtr0331651\"; transcript_symbol \"Gr21a-RB\";\n",
    "2L\tFlyBase\texon\t782238\t782441\t.\t+\t.\tgene_id \"FBgn0041250\"; gene_symbol \"Gr21a\"; transcript_id \"FBtr0331651\"; transcript_symbol \"Gr21a-RB\";\n",
    "2L\tFlyBase\texon\t782495\t782885\t.\t+\t.\tgene_id \"FBgn0041250\"; gene_symbol \"Gr21a\"; transcript_id \"FBtr0331651\"; transcript_symbol \"Gr21a-RB\";\n",
    "2L\tFlyBase\tstart_codon\t781297\t781299\t.\t+\t0\tgene_id \"FBgn0041250\"; gene_symbol \"Gr21a\"; transcript_id \"FBtr0331651\"; transcript_symbol \"Gr21a-RB\";\n",
    "2L\tFlyBase\tCDS\t781297\t782048\t.\t+\t0\tgene_id \"FBgn0041250\"; gene_symbol \"Gr21a\"; transcript_id \"FBtr0331651\"; transcript_symbol \"Gr21a-RB\";\n",
    "2L\tFlyBase\tCDS\t782124\t782181\t.\t+\t1\tgene_id \"FBgn0041250\"; gene_symbol \"Gr21a\"; transcript_id \"FBtr0331651\"; transcript_symbol \"Gr21a-RB\";\n",
    "2L\tFlyBase\tCDS\t782238\t782441\t.\t+\t0\tgene_id \"FBgn0041250\"; gene_symbol \"Gr21a\"; transcript_id \"FBtr0331651\"; transcript_symbol \"Gr21a-RB\";\n",
    "2L\tFlyBase\tCDS\t782495\t782821\t.\t+\t0\tgene_id \"FBgn0041250\"; gene_symbol \"Gr21a\"; transcript_id \"FBtr0331651\"; transcript_symbol \"Gr21a-RB\";\n",
    "2L\tFlyBase\tstop_codon\t782822\t782824\t.\t+\t0\tgene_id \"FBgn0041250\"; gene_symbol \"Gr21a\"; transcript_id \"FBtr0331651\"; transcript_symbol \"Gr21a-RB\";\n",
    "2L\tFlyBase\t3UTR\t782825\t782885\t.\t+\t.\tgene_id \"FBgn0041250\"; gene_symbol \"Gr21a\"; transcript_id \"FBtr0331651\"; transcript_symbol \"Gr21a-RB\";\n",
    "```\n",
    "\n",
    "\n",
    "We can also see how much content is in the file `wc` give a word count and with its `-l` parameter gives us the number of lines.\n",
    "```\n",
    "[brownsarahm@seawulf ~]$ wc -l dmel-all-r6.19.gtf\n",
    "```\n",
    "\n",
    "```\n",
    "542048 dmel-all-r6.19.gtf\n",
    "```\n",
    "Over five hundred forty thousand lines is a lot.\n",
    "\n",
    "\n",
    "How can we get the number of lines in each of the `.fastq` files?\n",
    "\n",
    "```\n",
    "[brownsarahm@seawulf ~]$ wc -l *.fastw\n",
    "wc: *.fastw: No such file or directory\n",
    "```\n",
    "note that in my typo, it tells me no files matched my pattern.\n",
    "\n",
    "```\n",
    "[brownsarahm@seawulf ~]$ wc -l *.fastq\n",
    "   20000 SRR307023_1.fastq\n",
    "   20000 SRR307023_2.fastq\n",
    "   20000 SRR307024_1.fastq\n",
    "   20000 SRR307024_2.fastq\n",
    "   20000 SRR307025_1.fastq\n",
    "   20000 SRR307025_2.fastq\n",
    "   20000 SRR307026_1.fastq\n",
    "   20000 SRR307026_2.fastq\n",
    "   20000 SRR307027_1.fastq\n",
    "   20000 SRR307027_2.fastq\n",
    "   20000 SRR307028_1.fastq\n",
    "   20000 SRR307028_2.fastq\n",
    "   20000 SRR307029_1.fastq\n",
    "   20000 SRR307029_2.fastq\n",
    "   20000 SRR307030_1.fastq\n",
    "   20000 SRR307030_2.fastq\n",
    "  320000 total\n",
    "```\n",
    "\n",
    "when it does work, we also get the total.\n",
    "\n",
    "We can use redirects as before to save these to a file:\n",
    "```\n",
    "[brownsarahm@seawulf ~]$ wc -l *.fastq > linecounts.txt\n",
    "```\n",
    "\n",
    "```\n",
    "[brownsarahm@seawulf ~]$ cat linecounts.txt\n",
    "   20000 SRR307023_1.fastq\n",
    "   20000 SRR307023_2.fastq\n",
    "   20000 SRR307024_1.fastq\n",
    "   20000 SRR307024_2.fastq\n",
    "   20000 SRR307025_1.fastq\n",
    "   20000 SRR307025_2.fastq\n",
    "   20000 SRR307026_1.fastq\n",
    "   20000 SRR307026_2.fastq\n",
    "   20000 SRR307027_1.fastq\n",
    "   20000 SRR307027_2.fastq\n",
    "   20000 SRR307028_1.fastq\n",
    "   20000 SRR307028_2.fastq\n",
    "   20000 SRR307029_1.fastq\n",
    "   20000 SRR307029_2.fastq\n",
    "   20000 SRR307030_1.fastq\n",
    "   20000 SRR307030_2.fastq\n",
    "  320000 total\n",
    "```\n",
    "\n",
    "\n",
    "We can also search files, without loading them all into memory or displaying them, with `grep`:\n",
    "```\n",
    "[brownsarahm@seawulf ~]$ grep Act5c dmel-all-r6.19.gtf\n",
    "```\n",
    "\n",
    "\n",
    "```\n",
    "[brownsarahm@seawulf ~]$ grep mRNA dmel-all-r6.19.gtf\n",
    "```\n",
    "\n",
    "this output a lot, so the output is truncated here\n",
    "```\n",
    "X\tFlyBase\tmRNA\t19961689\t19968479\t.\t+\t.\tgene_id \"FBgn0031081\"; gene_symbol \"Nep3\"; transcript_id \"FBtr0070000\"; transcript_symbol \"Nep3-RA\";\n",
    "2L\tFlyBase\tmRNA\t781276\t782885\t.\t+\t.\tgene_id \"FBgn0041250\"; gene_symbol \"Gr21a\"; transcript_id \"FBtr0331651\"; transcript_symbol \"Gr21a-RB\";\n",
    "```\n",
    "\n",
    "and we can combine `grep` with `wc` to count occurences.\n",
    "\n",
    "```\n",
    "[brownsarahm@seawulf ~]$ grep mRNA dmel-all-r6.19.gtf | wc -l\n",
    "34025\n",
    "```\n",
    "\n",
    "\n",
    "##  File permissions\n",
    "\n",
    "Let's make a small script, recalling what we have learned so far:\n",
    "```\n",
    "[brownsarahm@seawulf ~]$ echo \"echo 'script works'\" >> demo.sh\n",
    "```\n",
    "\n",
    "We can confirm that the script looks like a we expected\n",
    "```\n",
    "[brownsarahm@seawulf ~]$ cat demo.sh\n",
    "echo 'script works'\n",
    "```\n",
    "\n",
    "One thing we could do is to run the script using `./`\n",
    "```\n",
    "[brownsarahm@seawulf ~]$ ./demo.sh\n",
    "```\n",
    "\n",
    "but we get a permission denied error\n",
    "```\n",
    "-bash: ./demo.sh: Permission denied\n",
    "```\n",
    "\n",
    "By default, files have different types of permissions: read, write, and execute for different users that can access them. To view the permissions, we can use the `-l` option of `ls`.\n",
    "\n",
    "```\n",
    "[brownsarahm@seawulf ~]$ ls -l\n",
    "total 138452\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392 12534006 Apr 18  2021 bash-lesson.tar.gz\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392       20 Mar  8 13:12 demo.sh\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392 77426528 Jan 16  2018 dmel-all-r6.19.gtf\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392   721242 Jan 25  2016 dmel_unique_protein_isoforms_fb_2016_01.tsv\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392 25056938 Jan 25  2016 gene_association.fb\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392      447 Mar  8 13:07 linecounts.txt\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625262 Jan 25  2016 SRR307023_1.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625262 Jan 25  2016 SRR307023_2.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625376 Jan 25  2016 SRR307024_1.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625376 Jan 25  2016 SRR307024_2.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625286 Jan 25  2016 SRR307025_1.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625286 Jan 25  2016 SRR307025_2.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625302 Jan 25  2016 SRR307026_1.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625302 Jan 25  2016 SRR307026_2.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625312 Jan 25  2016 SRR307027_1.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625312 Jan 25  2016 SRR307027_2.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625338 Jan 25  2016 SRR307028_1.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625338 Jan 25  2016 SRR307028_2.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625390 Jan 25  2016 SRR307029_1.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625390 Jan 25  2016 SRR307029_2.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625318 Jan 25  2016 SRR307030_1.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625318 Jan 25  2016 SRR307030_2.fastq\n",
    "```\n",
    "\n",
    "For each file we get 10 characters in the first column that describe the permissions.  The 3rd column is the username of the owner, the fourth is the group, then size date revised and the file name.\n",
    "\n",
    "We are most interested in the 10 character permissions. The fist column indicates if any are directories with a `d` or a `-` for files. We have no directories, but we can create one to see this.\n",
    "\n",
    "```\n",
    "[brownsarahm@seawulf ~]$ mkdir results\n",
    "```\n",
    "\n",
    "```\n",
    "[brownsarahm@seawulf ~]$ ls -l\n",
    "total 138452\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392 12534006 Apr 18  2021 bash-lesson.tar.gz\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392       20 Mar  8 13:12 demo.sh\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392 77426528 Jan 16  2018 dmel-all-r6.19.gtf\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392   721242 Jan 25  2016 dmel_unique_protein_isoforms_fb_2016_01.tsv\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392 25056938 Jan 25  2016 gene_association.fb\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392      447 Mar  8 13:07 linecounts.txt\n",
    "**drwxr-xr-x. 2 brownsarahm spring2022-csc392       10 Mar  8 13:16 results**\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625262 Jan 25  2016 SRR307023_1.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625262 Jan 25  2016 SRR307023_2.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625376 Jan 25  2016 SRR307024_1.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625376 Jan 25  2016 SRR307024_2.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625286 Jan 25  2016 SRR307025_1.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625286 Jan 25  2016 SRR307025_2.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625302 Jan 25  2016 SRR307026_1.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625302 Jan 25  2016 SRR307026_2.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625312 Jan 25  2016 SRR307027_1.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625312 Jan 25  2016 SRR307027_2.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625338 Jan 25  2016 SRR307028_1.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625338 Jan 25  2016 SRR307028_2.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625390 Jan 25  2016 SRR307029_1.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625390 Jan 25  2016 SRR307029_2.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625318 Jan 25  2016 SRR307030_1.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625318 Jan 25  2016 SRR307030_2.fastq\n",
    "```\n",
    "\n",
    "We can see in the bold line, that the first character is a d.\n",
    "\n",
    "The next nine characters indicate permission to **R**ead, **W**rite, and e**X**ecute a file. With either the letter or a `-` for permissions not granted, they appear in three groups of three, three characters each for owner, group, anyone with access.\n",
    "\n",
    "If we want to run the file, we *can* instead use `bash` directly, but this is limited relative to calling our script in other ways.\n",
    "\n",
    "```\n",
    "[brownsarahm@seawulf ~]$ bash demo.sh\n",
    "script works\n",
    "```\n",
    "\n",
    "Instead, to add execute permission, we can use `chmod`\n",
    "\n",
    "```\n",
    "[brownsarahm@seawulf ~]$ chmod +x demo.sh\n",
    "```\n",
    "\n",
    "\n",
    "\n",
    "```\n",
    "[brownsarahm@seawulf ~]$ ls -l\n",
    "total 138452\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392 12534006 Apr 18  2021 bash-lesson.tar.gz\n",
    "-rwxr-xr-x. 1 brownsarahm spring2022-csc392       20 Mar  8 13:12 demo.sh\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392 77426528 Jan 16  2018 dmel-all-r6.19.gtf\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392   721242 Jan 25  2016 dmel_unique_protein_isoforms_fb_2016_01.tsv\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392 25056938 Jan 25  2016 gene_association.fb\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392      447 Mar  8 13:07 linecounts.txt\n",
    "drwxr-xr-x. 2 brownsarahm spring2022-csc392       10 Mar  8 13:16 results\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625262 Jan 25  2016 SRR307023_1.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625262 Jan 25  2016 SRR307023_2.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625376 Jan 25  2016 SRR307024_1.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625376 Jan 25  2016 SRR307024_2.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625286 Jan 25  2016 SRR307025_1.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625286 Jan 25  2016 SRR307025_2.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625302 Jan 25  2016 SRR307026_1.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625302 Jan 25  2016 SRR307026_2.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625312 Jan 25  2016 SRR307027_1.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625312 Jan 25  2016 SRR307027_2.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625338 Jan 25  2016 SRR307028_1.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625338 Jan 25  2016 SRR307028_2.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625390 Jan 25  2016 SRR307029_1.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625390 Jan 25  2016 SRR307029_2.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625318 Jan 25  2016 SRR307030_1.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625318 Jan 25  2016 SRR307030_2.fastq\n",
    "```\n",
    "\n",
    "```\n",
    "[brownsarahm@seawulf ~]$ ./demo.sh\n",
    "script works\n",
    "```\n",
    "\n",
    "```\n",
    "[brownsarahm@seawulf ~]$ nano demo.sh\n",
    "```\n",
    "\n",
    "We can add a bit more to our script to make it more interesting\n",
    "```\n",
    "for VAR in *.gz\n",
    "do\n",
    "  echo $VAR\n",
    "done\n",
    "\n",
    "\n",
    "echo 'script works'\n",
    "```\n",
    "\n",
    "and note that tht does not change the permission.\n",
    "```\n",
    "[brownsarahm@seawulf ~]$ ls -l\n",
    "total 138452\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392 12534006 Apr 18  2021 bash-lesson.tar.gz\n",
    "-rwxr-xr-x. 1 brownsarahm spring2022-csc392       60 Mar  8 13:27 demo.sh\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392 77426528 Jan 16  2018 dmel-all-r6.19.gtf\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392   721242 Jan 25  2016 dmel_unique_protein_isoforms_fb_2016_01.tsv\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392 25056938 Jan 25  2016 gene_association.fb\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392      447 Mar  8 13:07 linecounts.txt\n",
    "drwxr-xr-x. 2 brownsarahm spring2022-csc392       10 Mar  8 13:16 results\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625262 Jan 25  2016 SRR307023_1.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625262 Jan 25  2016 SRR307023_2.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625376 Jan 25  2016 SRR307024_1.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625376 Jan 25  2016 SRR307024_2.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625286 Jan 25  2016 SRR307025_1.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625286 Jan 25  2016 SRR307025_2.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625302 Jan 25  2016 SRR307026_1.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625302 Jan 25  2016 SRR307026_2.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625312 Jan 25  2016 SRR307027_1.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625312 Jan 25  2016 SRR307027_2.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625338 Jan 25  2016 SRR307028_1.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625338 Jan 25  2016 SRR307028_2.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625390 Jan 25  2016 SRR307029_1.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625390 Jan 25  2016 SRR307029_2.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625318 Jan 25  2016 SRR307030_1.fastq\n",
    "-rw-r--r--. 1 brownsarahm spring2022-csc392  1625318 Jan 25  2016 SRR307030_2.fastq\n",
    "```\n",
    "\n",
    "\n",
    "##  Making a batch file\n",
    "\n",
    "```\n",
    "[brownsarahm@seawulf ~]$ nano my_job.sh\n",
    "```\n",
    "\n",
    "```\n",
    "[brownsarahm@seawulf ~]$ sbatch my_job.sh\n",
    "Submitted batch job 23950\n",
    "```\n",
    "\n",
    "```\n",
    "[brownsarahm@seawulf ~]$ squeue --job 23950\n",
    "             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)\n",
    "```\n",
    "\n",
    "## Interactive mode\n",
    "\n",
    "So far we have worked on the login node which has very little computational power and submitted a job to a compute node through the scheduler for it to run asynchronously.  We can also use a compute node interactively if we have a computationally intensive task that we need to work with in real time, instead of it being fully automated.  \n",
    "\n",
    "We'll talk after the break about how this might look, but for example if our `grep` commands needed more power where we need the resutl back right away in order to then decide what the next calculation could be.\n",
    "\n",
    "\n",
    "We can request a node to be allocated to our user and have a direct `ssh` connection to compute node using `interactive`.  Interactive by default calls a specific script that is configured by the system administrators for the particular cluster that you are logged into.  \n",
    "\n",
    "For seawulf, [interactive](https://web.uri.edu/hpc-research-computing/using-seawulf/#interactive) opens a default of 8 hgours, but you can use parameters to request more or less time.\n",
    "```\n",
    "[brownsarahm@seawulf ~]$ interactive\n",
    "```\n",
    "\n",
    "When we run this interactive submits a \"job\" to the schedule, but instead of then passing a script to the compute node it connects us to that node via ssh\n",
    "```\n",
    "salloc: Granted job allocation 23963\n",
    "salloc: Waiting for resource configuration\n",
    "salloc: Nodes n005 are ready for job\n",
    "/usr/bin/id: cannot find name for user ID 1168\n",
    "/usr/bin/id: cannot find name for group ID 1111\n",
    "/usr/bin/id: cannot find name for user ID 1168\n",
    "```\n",
    "\n",
    "\n",
    "We can see that we are connected to a system with more resources using this:\n",
    "```\n",
    "[I have no name!@n005 ~]$ lshw\n",
    "WARNING: you should run this program as super-user.\n",
    "          configuration: driver=system\n",
    "WARNING: output may be incomplete or inaccurate, you should run this program as super-user.\n",
    "```\n",
    "\n",
    "I truncataed the long output.\n",
    "\n",
    "We can then exit the compute node and go back to the login node with `exit`\n",
    "\n",
    "```\n",
    "[I have no name!@n005 ~]$ exit\n",
    "```\n",
    "\n",
    "\n",
    "\n",
    "```\n",
    "logout\n",
    "salloc: Relinquishing job allocation 23963\n",
    "```\n",
    "Note tht it says the job has ended.\n",
    "\n",
    "## Logging off\n",
    "\n",
    "You can log off with `exit` it is important to do so to not use resources you do not need. However, note tht if your laptop loses internet connection your session will also be disconnected.  This is another advantage of the scheduler, you don't need to maintain a connection for it to run your code.\n",
    "\n",
    "\n",
    "```\n",
    "[brownsarahm@seawulf ~]$ exit\n",
    "```\n",
    "\n",
    "```\n",
    "logout\n",
    "Connection to seawulf.uri.edu closed.\n",
    "```\n",
    "\n",
    "Note that when you are back to your computer the prompt changes. And that using the up arrow, the last command your computer's shell knows is the `ssh` line.\n",
    "```\n",
    "brownsarahm@~ $ ssh -l brownsarahm seawulf.uri.edu\n",
    "```\n",
    "\n",
    "(class14prep)=\n",
    "## Prepare for next class\n",
    "````{margin}\n",
    "```{note}\n",
    "The text in () in these sections is an explanation of *why* that task is assigned\n",
    "```\n",
    "````"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0df364df",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "1. Review the notes (to reinforce ideas and improve your memory of them)\n",
       "1. **priority** add a `## summary` section to your IDE notes where you summarize what you observed in your IDE usage and what questions or what features (we will work with this after break and I want you to prepare before break)\n",
       "1. Modify the sbatch job we made in class so that you get files back for any errors and output from the script. [see the options](https://web.uri.edu/hpc-research-computing/using-seawulf/#sbatch) (to review/practice what we did)\n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(Markdown(prepare))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e2c7c07",
   "metadata": {},
   "source": [
    "(class14practice)=\n",
    "## More Practice\n",
    "\n",
    "````{margin}\n",
    "```{admonition} Try it Yourself\n",
    "use [scp](https://man7.org/linux/man-pages/man1/scp.1.html) to download the files generated to your local computer and add them to your kwl chart.  (a possible extenstion idea to practice reading the official documentation to learn a new tool that is similar to ones you have seen before)\n",
    "```\n",
    "````"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "83046399",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "1. **priority** write a script that outputs the first 3 lines of each `.fastq` line and sbatch script that runs it and saves the output to files and e-mails you when it is done. Submit the job. (to reinforce /practice what we saw in class)\n",
       "1. File permissions are represented numerically often in octal, by transforming the permissions for each level (owner, group, all) as binary into a number. Add `octal.md` to your KWL repo and answer the following. Try to think through it on your own, but you may look up the answers, as long as you link to (or ideally cite using jupyterbook citations) a high quality source.\n",
       "    ```\n",
       "    1. Transform the permissions [`r--`, `rw-`, `rwx`] to octal, by treating it as three bits.\n",
       "    1. Transform the permission we changed our script to `rwxr-xr-x` to octal.\n",
       "    1. Which of the following would prevent a file from being edited by other people 777 or 755?\n",
       "    ```\n",
       "1. Answer the following in `hpc.md` of your KWL repo:  (to think about how the design of the system we used in class impacts programming and connect it to other ideas taught in CS)\n",
       "    ```\n",
       "    1. What kinds of things would your code need to do if you were going to run it on an HPC system?\n",
       "    1. What Sbatch options seem the most helpful?\n",
       "    1. How might you go about setting the time limits for a script? How could you estimate how long it will take?\n",
       "    ```\n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(Markdown(practice))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66f305aa",
   "metadata": {},
   "source": [
    "## Questions After Class\n",
    "\n",
    "\n",
    "### When is sbatch used?\n",
    "```{toggle}\n",
    "Sbatch is one type of scheduler, that manages the resource of a high performance computing system\n",
    "```\n",
    "\n",
    "### will sbatch kick out your job after a certain amount of time? what if your job was indefinite like hosting a website.\n",
    "```{toggle}\n",
    "High performance computing systems, like the cluster that we logged into in class are ued for computationally expensive programs that might require more RAM and/or more CPUs in order to run and/or to run for a longer time than your laptop makes sense to leave it in one place.  \n",
    "\n",
    "For example, this is often used in research data analysis, like the genomics data we used today or for training machine learning models.\n",
    "\n",
    "HPCs are a specifc type of remote resource, for indefinite jobs, we use different types of servers.  However, `ssh` can still be used to log into those sytems.\n",
    "```"
   ]
  }
 ],
 "metadata": {
  "jupytext": {
   "text_representation": {
    "extension": ".md",
    "format_name": "myst",
    "format_version": 0.13,
    "jupytext_version": "1.10.3"
   }
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  },
  "source_map": [
   15,
   23,
   598,
   602,
   615,
   618
  ]
 },
 "nbformat": 4,
 "nbformat_minor": 5
}